crates/nu-command/src/filters/join.rs

```diff
@@ -0,0 +1,411 @@
+                format!(
+                    "({:?}, {:?}, {:?} {:?})",
+                    collected_input,
+                    table_2.get_type(),
+                    l_on.get_type(),
+                    r_on.get_type(),
+                ),
+                span,
+                span,
+            )),
+        }
+    }
+
+    fn examples(&self) -> Vec<Example> {
+        vec![Example {
+            description: "Join two tables",
+            example: "[{a: 1 b: 2}] | join [{a: 1 c: 3}] a",
+            result: Some(Value::List {
+                vals: vec![Value::Record {
+                    cols: vec!["a".into(), "b".into(), "c".into()],
+                    vals: vec![
+                        Value::Int {
+                            val: 1,
+                            span: Span::test_data(),
+                        },
+                        Value::Int {
+                            val: 2,
+                            span: Span::test_data(),
+                        },
+                        Value::Int {
+                            val: 3,
+                            span: Span::test_data(),
+                        },
+                    ],
+                    span: Span::test_data(),
+                }],
+                span: Span::test_data(),
+            }),
+        }]
+    }
+}
+
+fn join_type(call: &Call) -> Result<JoinType, nu_protocol::ShellError> {
+    match (
+        call.has_flag("inner"),
+        call.has_flag("left"),
+        call.has_flag("right"),
+        call.has_flag("outer"),
+    ) {
+        (_, false, false, false) => Ok(JoinType::Inner),
+        (false, true, false, false) => Ok(JoinType::Left),
+        (false, false, true, false) => Ok(JoinType::Right),
+        (false, false, false, true) => Ok(JoinType::Outer),
+        _ => Err(ShellError::UnsupportedInput(
+            "Choose one of: --inner, --left, --right, --outer".into(),
+            "".into(),
+            call.head,
+            call.head,
+        )),
+    }
+}
+
+fn join(
+    left: &Vec<Value>,
+    right: &Vec<Value>,
+    l_on: &str,
+    r_on: &str,
+    join_type: JoinType,
+    span: Span,
+) -> Value {
+    // Inner / Right Join
+    // ------------------
+    // Make look-up table from rows on left
+    // For each row r on right:
+    //    If any matching rows on left:
+    //        For each matching row l on left:
+    //            Emit (l, r)
+    //    Else if RightJoin:
+    //        Emit (null, r)
+
+    // Left Join
+    // ----------
+    // Make look-up table from rows on right
+    // For each row l on left:
+    //    If any matching rows on right:
+    //        For each matching row r on right:
+    //            Emit (l, r)
+    //    Else:
+    //        Emit (l, null)
+
+    // Outer Join
+    // ----------
+    // Perform Left Join procedure
+    // Perform Right Join procedure, but excluding rows in Inner Join
+
+    let config = Config::default();
```

Comment:
Guess it's OK to use the default config since we're using this for a hash key not display purposes 👍